### PR TITLE
Review RegionYolo class for shape inference aspects

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Review PartialShape and labels propagation.
 - Review StaticShape propagation.
 - Review template implementation of `shape_infer` function.

### Tickets:
 - 97247
 - 97291
 - 97340
 - 97389

### CODE REVIEW

File: src/plugins/intel_cpu/tests/unit/shape_inference_test/region_yolo_shape_inference_test.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+TEST_F(StaticShapeRegionYoloTest, data_shape_not_compatible_rank_4) {
+    const auto data = std::make_shared<op::v0::Parameter>(element::f32, PartialShape::dynamic());
+    op = make_op(data, 5, 4, 20, false, std::vector<int64_t>{0, 1}, 1, 3);
+
+    OV_EXPECT_THROW(shape_inference(op.get(), ShapeVector{{2, 20, 12, 24, 1}}, output_shapes),
+                    NodeValidationFailure,
+                    HasSubstr("Input must be a tensor of rank 4, but got"));
+}
+
+// TEST_F(StaticShapeRegionYoloTest, h_dim_not_div_by_stride) {
```
Comment (jane-intel) [inline src/plugins/intel_cpu/tests/unit/shape_inference_test/region_yolo_shape_inference_test.cpp:62]: Is this PR ready for review?

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/tests/unit/shape_inference_test/region_yolo_shape_inference_test.cpp (hunk 1 above)
Comment (praasz) [reply src/plugins/intel_cpu/tests/unit/shape_inference_test/region_yolo_shape_inference_test.cpp:62]: Yes, it is leftover just missed will fix it soon.

File: src/core/shape_inference/include/region_yolo_shape_inference.hpp [hunk 2]
```diff
@@ -4,57 +4,58 @@
...
-                flat_dim *= input_shape[i];
-            }
-            output_shape.push_back(flat_dim);
-            for (size_t i = end_axis + 1; i < input_shape.size(); i++) {
-                output_shape.push_back(input_shape[i]);
+
+            std::copy_n(input_shape.cbegin(), axis, std::back_inserter(output_shape));
+
+            if (end_axis > axis) {
+                output_shape.push_back(input_shape[axis]);
```
Comment (mitruska) [inline src/core/shape_inference/include/region_yolo_shape_inference.hpp:46]: What if `end_axis == axis`?
Shouldn't the `input_shape[axis]` be added to the output_shape?
```suggestion
            std::copy_n(input_shape.cbegin(), axis, std::back_inserter(output_shape));

            if (end_axis >= axis) {
                output_shape.push_back(input_shape[axis]);
```

Comment (praasz) [reply src/core/shape_inference/include/region_yolo_shape_inference.hpp]: Yes, but intension was different, re-factored.

**Review by mitruska — APPROVED** [review]

**Review by jane-intel — APPROVED** [review]

**Pull request merged**